File: typescript/end-to-end-applications/chat-bot/src/taskmanager.ts
import * as restate from "@restatedev/restate-sdk"
import type { ChatSession } from "./chat"

// ----------------------------------------------------------------------------
//  The Task Manager has the map of available task workflows.
//  It maintains the mapping from task_type (name of the task type) to the
//  implementing workflow service, and has the utilities to start, cancel,
//  and query them.
// ----------------------------------------------------------------------------

// ------------------ defining new types of task workflows --------------------

export type TaskWorkflow<P = unknown> = {

    run: (ctx: restate.WorkflowContext, params: P) => Promise<string>

    cancel: (ctx: restate.WorkflowSharedContext) => Promise<void>,

    currentStatus: (ctx: restate.WorkflowSharedContext) => Promise<unknown>
}

export type TaskSpec<P> = {
    taskTypeName: string,
    taskWorkflow: restate.WorkflowDefinition<string, TaskWorkflow<P>>,
    paramsParser: (taskName: string, params: object) => P,
}

export function registerTaskWorkflow<P>(task: TaskSpec<P>) {
    availableTaskTypes.set(task.taskTypeName, task);   
}

const availableTaskTypes: Map<string, TaskSpec<unknown>> = new Map();

// ----------------- start / cancel / query task workflows --------------------

export type TaskOpts = {
    name: string,
    workflowName: string,
    params: object
}

export type TaskResult = { taskName: string, result: string }

export async function startTask<P = unknown>(
        ctx: restate.Context,
        channelForResult: string,
        taskOps: TaskOpts): Promise<string> {

    const task = availableTaskTypes.get(taskOps.workflowName) as TaskSpec<P> | undefined;
    if (!task) {
        throw new Error("Unknown task type: " + taskOps.workflowName);
    }

    const workflowParams = task.paramsParser(taskOps.name, taskOps.params)
    const workflowId = ctx.rand.uuidv4();

    ctx.serviceSendClient(workflowInvoker).invoke({
            taskName: taskOps.name,
            workflowServiceName: task.taskWorkflow.name,
            workflowParams,
            workflowId,
            channelForResult
        });
    
    return workflowId;
}

export async function cancelTask(
        ctx: restate.Context,
        workflowName: string,
        workflowId: string): Promise<void> {

    const task = availableTaskTypes.get(workflowName);
    if (!task) {
        throw new Error("Unknown task type: " + workflowName);
    }

    await ctx.workflowClient(task.taskWorkflow, workflowId).cancel();
}

export async function getTaskStatus(
        ctx: restate.Context,
        workflowName: string,
        workflowId: string): Promise<unknown> {

    const task = availableTaskTypes.get(workflowName);
    if (!task) {
        throw new Error("Unknown task type: " + workflowName);
    }

    const response = ctx.workflowClient(task.taskWorkflow, workflowId).currentStatus();
    return response;
}


// ----------------------------------------------------------------------------
//  Utility durable function that awaits the workflow result and forwards
//  it to the chat session
// ----------------------------------------------------------------------------

export const workflowInvoker = restate.service({
    name: "workflowInvoker",
    handlers: {
        invoke: async (
                ctx: restate.Context,
                opts: {
                    workflowServiceName: string,
                    workflowId: string,
                    workflowParams: unknown,
                    taskName: string,
                    channelForResult: string
                }) => {

            const taskWorkflowApi: restate.WorkflowDefinition<string, TaskWorkflow<unknown>> = { name: opts.workflowServiceName };
            let response: TaskResult;
            try {
                const result = await ctx.workflowClient(taskWorkflowApi, opts.workflowId).run(opts.workflowParams);
                response = { taskName: opts.taskName, result };
            } catch (err: any) {
                response = { taskName: opts.taskName, result: "Task failed: " +err.message }
            }

            ctx.objectSendClient<ChatSession>({ name: "chatSession" }, opts.channelForResult)
                .taskDone(response);
        }
    }
})

File: python/end-to-end-applications/chat-bot/chatbot/taskmanager.py
"""
  The Task Manager has the map of available task workflows.
  It maintains the mapping from task_type (name of the task type) to the
  implementing workflow service, and has the utilities to start, cancel,
  and query them.
"""
import logging
import uuid
from typing import Dict, Any

from restate import Context, Service
from restate.exceptions import TerminalError

from chatbot import chat_session
from chatbot.tasks.task_workflow import TaskSpec
from chatbot.utils.types import TaskResult, TaskOpts

from chatbot.tasks.flight_price_watcher import flightTask
from chatbot.tasks.reminder_service import reminderTask


TASK_TYPES: Dict[str, TaskSpec] = {
    reminderTask.task_type_name: reminderTask,
    flightTask.task_type_name: flightTask
    }

# ----------------- start / cancel / query task workflows --------------------

async def start_task(ctx: Context, channel_for_result: str, task_opts: TaskOpts) -> str:
    task = TASK_TYPES.get(task_opts.workflow_name)
    if not task:
        raise ValueError(f"Unknown task type: {task_opts.workflow_name}")

    workflow_params = task.params_parser(task_opts.name, task_opts.params)
    workflow_id = await ctx.run("workflow_id", lambda: str(uuid.uuid4()))

    ctx.service_send(invoke_workflow, {
        'task_name': task_opts.name,
        'workflow_service_name': task_opts.workflow_name,
        'workflow_params': workflow_params,
        'workflow_id': workflow_id,
        'channel_for_result': channel_for_result
    })

    return workflow_id


async def cancel_task(ctx: Context, workflow_name: str, workflow_id: str) -> None:
    task = TASK_TYPES.get(workflow_name)
    if not task:
        raise ValueError(f"Can't cancel task type for workflow ID {workflow_id} - Unknown task type: {workflow_name}")

    await ctx.workflow_call(task.task_workflow.cancel, workflow_id, None)


async def get_task_status(ctx: Context, workflow_name: str, workflow_id: str) -> Any:
    task = TASK_TYPES.get(workflow_name)
    if not task:
        raise ValueError(f"Can't get task status for workflow ID {workflow_id} - Unknown task type: {workflow_name}")

    return await ctx.workflow_call(task.task_workflow.current_status, workflow_id, None)


workflow_invoker = Service("WorkflowInvoker")

@workflow_invoker.handler("invokeWorkflow")
async def invoke_workflow(ctx: Context, opts: dict) -> None:
    logging.info("Invoking workflow: %s", opts)
    task = TASK_TYPES.get(opts["workflow_service_name"])
    if not task:
        logging.error("Unknown task type: %s", opts["workflow_service_name"])
        raise TerminalError(f"Unknown task type: {opts['workflow_service_name']}")
    response: TaskResult
    try:
        result = await ctx.workflow_call(task.task_workflow.run, opts['workflow_id'], opts['workflow_params'])
        response = TaskResult(task_name=opts['task_name'], result=result)
    except Exception as e:
        response = TaskResult(task_name=opts['task_name'], result=f"Task failed: {str(e)}")

    ctx.object_send(chat_session.on_task_done, opts['channel_for_result'], response)
